File: src/main/java/io/lsdconsulting/lsd/distributed/interceptor/config/RestTemplateInterceptorConfig.java
package io.lsdconsulting.lsd.distributed.interceptor.config;

import io.lsdconsulting.lsd.distributed.interceptor.captor.http.RequestCaptor;
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.ResponseCaptor;
import io.lsdconsulting.lsd.distributed.interceptor.interceptor.LsdRestTemplateCustomizer;
import io.lsdconsulting.lsd.distributed.interceptor.interceptor.LsdRestTemplateInterceptor;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.web.client.RestTemplate;

@Slf4j
@Configuration
@ConditionalOnProperty(name = "lsd.dist.db.connectionString")
@ConditionalOnClass(RestTemplate.class)
@RequiredArgsConstructor
public class RestTemplateInterceptorConfig {

    @Bean
    public ClientHttpRequestInterceptor lsdRestTemplateInterceptor(final RequestCaptor requestCaptor, final ResponseCaptor responseCaptor) {
        return new LsdRestTemplateInterceptor(requestCaptor, responseCaptor);
    }

    @Bean
    public LsdRestTemplateCustomizer lsdRestTemplateCustomizer(final ClientHttpRequestInterceptor lsdRestTemplateInterceptor) {
        return new LsdRestTemplateCustomizer(lsdRestTemplateInterceptor);
    }
}

File: spring-web/src/main/kotlin/io/lsdconsulting/lsd/distributed/interceptor/config/RestTemplateInterceptorConfig.kt
package io.lsdconsulting.lsd.distributed.interceptor.config

import io.lsdconsulting.lsd.distributed.interceptor.captor.RequestCaptor
import io.lsdconsulting.lsd.distributed.interceptor.captor.ResponseCaptor
import io.lsdconsulting.lsd.distributed.interceptor.interceptor.LsdRestTemplateCustomizer
import io.lsdconsulting.lsd.distributed.interceptor.interceptor.LsdRestTemplateInterceptor
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.web.client.RestTemplate

@Configuration
@ConditionalOnProperty(name = ["lsd.dist.connectionString"])
@ConditionalOnClass(RestTemplate::class)
class RestTemplateInterceptorConfig {
    @Bean
    fun lsdRestTemplateInterceptor(
        requestCaptor: RequestCaptor,
        responseCaptor: ResponseCaptor
    ): ClientHttpRequestInterceptor {
        return LsdRestTemplateInterceptor(requestCaptor, responseCaptor)
    }

    @Bean
    fun lsdRestTemplateCustomizer(lsdRestTemplateInterceptor: ClientHttpRequestInterceptor?): LsdRestTemplateCustomizer {
        return LsdRestTemplateCustomizer(lsdRestTemplateInterceptor!!)
    }
}
